### DISCUSSION TITLE
Accessing from outside response "Bad Request-Invalid Hostname"

### DISCUSSION THREAD

**Discussion by Jycjmf** [dotnet/WatsonWebserver#61] [Q&A]
HI!
I want to access the WebServer from outside, but it response "Bad Request-Invalid Hostname"; I have read the ReadMe about "Accessing from Outside Localhost", but I cannot understand the actual mean of  " exact DNS hostname ". 
Here is my constructor, I just use localhost as hostname
```
    Server s = new Server("localhost", 2333, false, DefaultRoute);
    s.Start();
```
And here is my POST request, I added the headers named "Host" and the value is "localhost":
![image](https://user-images.githubusercontent.com/30173232/112746799-d8a56c00-8fe3-11eb-8bcf-7b183ea0d034.png)

I have already binding the URL like below:
![image](https://user-images.githubusercontent.com/30173232/112746878-81ec6200-8fe4-11eb-8635-c7984f549919.png)

And this is my public ip address : 39.99.135.10 , I just want to access the webserver from outer ip address, what should I do?
Thanks!

**Comment by jchristn** [comment]
Hi @Jycjmf

If you set your listener to ```localhost``` (i.e. ```Server s = new Server("localhost", 2333, false, DefaultRoute);```) it will not accept external connections.  Try setting the listener IP to either A) your public IP address or B) a wildcard ```*``` or ```+```.  Doing so will allow you to *listen* on multiple interfaces, but you'll have to run as administrator.

You shouldn't have to manually set the HOST header in Postman. It may automatically set it for you based on the URL. The HOST header has no bearing on connecting to your machine, but rather, what the underlying ```http.sys``` webserver does with the request once received (i.e. it consults your URL reservations)

So the issue with the HOST header is that these things have to match:

- HTTP URL reservation (you've set yours to ```localhost``` in the Command Prompt screenshot above)
- HTTP listener prefix (which you're setting to ```localhost``` in ```Server s = new Server...```)

If you try to access using a public IP, you'll run into two issues:

1) Listening on ```localhost``` means you can't accept external connections
2) When you connect via the public IP (```39.99.135.10```) the HOST header will be set to that IP address, and not ```localhost```, meaning the URL reservation won't match, meaning the client will be returned a 400 - Invalid Hostname.

Try listening on ```*``` or ```+```, or, only listening on ```39.99.135.10```, and then re-do your URL reservation to the same value (```*```, ```+```, or ```39.99.135.10```).  If you use a wildcard, you'll have to run as administrator.

Cheers,
Joel

**Reply by Jycjmf** [reply]
HI，@jchristn , I am so sorry for the delay in responding to you.
I have tried what you said above, but it cannot get any response in Postman.
![image](https://user-images.githubusercontent.com/30173232/112851439-8d6f8400-90dd-11eb-9aeb-b567aaf463c9.png)
I have changed the listener prefix to my public ip address:
![image](https://user-images.githubusercontent.com/30173232/112851517-a1b38100-90dd-11eb-8a71-a69014fdffd4.png)
And re-do the URL reservation to the same value:
![image](https://user-images.githubusercontent.com/30173232/112851643-c3ad0380-90dd-11eb-9075-b969085f9380.png)
But it still not work ,I am looking forward to your reply ,thanks!

**Reply by jchristn** [reply]
Hi @Jycjmf - is 39.99.135.10 the actual IP address on the network interface that you see when you do ipconfig on the machine?

You can also try setting the listener prefix and URL reservation both to + or * and running as administrator.

**Reply by Jycjmf** [reply]
Hi @jchristn ,It works fine when I change the listener prefix to my Primary Private IP Address, and Post request to my public IP Address
![image](https://user-images.githubusercontent.com/30173232/112927479-a14fd000-9147-11eb-90c5-2b78f68602ab.png)
![image](https://user-images.githubusercontent.com/30173232/112927793-22a76280-9148-11eb-9a58-9832d7d49a60.png)
![image](https://user-images.githubusercontent.com/30173232/112927538-c17f8f00-9147-11eb-8dd9-ee4296c9bd4e.png)

But the URL reservation seems useless in my demo, because I have deleted the URL reservation I added before, It still work fine.
![image](https://user-images.githubusercontent.com/30173232/112927714-04d9fd80-9148-11eb-9935-61d9b2377d5d.png)
![image](https://user-images.githubusercontent.com/30173232/112927744-0c99a200-9148-11eb-9b39-3148d7ff5ccd.png)
Anyway, thank you for your help , bro

**Reply by jchristn** [reply]
Hi @Jycjmf glad you got it working.  Cheers!

**Comment by jchristn** [comment]
Hi @Jycjmf wasn't sure if you saw my edits above.  I made them after the original response, so the update may not have come across your email.  Cheers, Joel
